### PR TITLE
[SYCLomatic] Support the migration of matmul & transform in cublasLt

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 1]
```diff
@@ -2467,8 +2467,618 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+  if (compute_desc->_epilogue != 1) {
+    throw std::runtime_error("Epilogue is not supported.");
+  }
+
+  // 3. only support row major and col major
+  if ((a_desc->_order != order_t::col && a_desc->_order != order_t::row) ||
+      (b_desc->_order != order_t::col && b_desc->_order != order_t::row) ||
+      (c_desc->_order != order_t::col && c_desc->_order != order_t::row) ||
+      (d_desc->_order != order_t::col && d_desc->_order != order_t::row)) {
+    throw std::runtime_error("Only support row majorand col major.");
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: We can transform other layouts to col major layout (allocate new memory). After computation, transform it back to the original memory. Maybe also emit some runtime messages.

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 2]
```diff
@@ -2467,8 +2467,618 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+            to_idx = c_ld * col_idx + row_idx;
+          }
+          c[to_idx] = (*alpha_data) * a[from_idx];
+        }
+      });
+    });
+  }
+}
+} // namespace detail
+void matrix_transform(transform_desc_ptr transform_desc, const void *alpha,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: We do not need transpose matrix currently (at least for bitsandbytes).
For this function, we need impl below conversations:
col_major <=> row_major,
col_major <=> COL32,
col_major <=> COL4_4R2_8C,
col_major <=> COL32_2R_4R4

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 3]
```diff
@@ -2467,8 +2467,618 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+            const void *beta, const void *c, matrix_layout_ptr c_desc, void *d,
+            matrix_layout_ptr d_desc, dpct::queue_ptr q_ptr) {
+  if (!q_ptr)
+    q_ptr = &get_default_queue();
+  if (compute_desc->_pointer_mode == pointer_mode_t::device_vector ||
+      compute_desc->_pointer_mode ==
+          pointer_mode_t::alpha_device_vector_beta_zero ||
+      compute_desc->_pointer_mode ==
+          pointer_mode_t::alpha_device_vector_beta_host) {
+    throw std::runtime_error("Unsupported pointer mode.");
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: Write SYCL kernel to impl vector alpha

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 4]
```diff
@@ -2467,8 +2467,618 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+
+  // 3. only support row major and col major
+  if ((a_desc->_order != order_t::col && a_desc->_order != order_t::row) ||
+      (b_desc->_order != order_t::col && b_desc->_order != order_t::row) ||
+      (c_desc->_order != order_t::col && c_desc->_order != order_t::row) ||
+      (d_desc->_order != order_t::col && d_desc->_order != order_t::row)) {
+    throw std::runtime_error("Only support row majorand col major.");
+  }
+
+  // TODO: process a_trans and b_trans
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: Consider if lda/ldb are not same as m/n/k.

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 5]
```diff
@@ -2467,8 +2467,618 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+  case library_data_t::real_uint8:
+    return dnnl::memory::data_type::u8;
+  default:
+    throw std::runtime_error(
+        "the input library_data_t type has no corresponding type in dnnl");
+  }
+}
+} // namespace detail
+
+void matmul(matmul_desc_ptr compute_desc, const void *alpha, const void *a,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: a/b: convert to col major layout, allocate new memory for col major layout d to save result, then do the calculation.
After calculation, convert col major layout d to col32 layout.

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 2 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 3 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 4 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: test passed

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 5 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 1 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: Calculation impl done. Runtime message/warning message not added yet.

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 6]
```diff
@@ -2467,8 +2467,737 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+                                                dst_md, matmul_attr);
+
+  // Create the primitive.
+  auto matmul_prim = dnnl::matmul(matmul_pd);
+
+  // Primitive execution: matrix multiplication with ReLU.
+  matmul_prim.execute(engine_stream, matmul_args);
+
+  // Wait for the computation to finalize.
+  engine_stream.wait();
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: How to remove wait() to make matmul async?

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 7]
```diff
@@ -2467,8 +2467,737 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+      detail::matrix_transform<int>(q_ptr, a_desc->_rows, a_desc->_cols,
+                                    a_desc->_ld, a_desc->_order, (const int *)a,
+                                    new_lda, order_t::col, (int *)new_a);
+    }
+  }
+  if (b_desc->_order != order_t::col) {
+    new_ldb = b_desc->_cols; // The input b is transpose::trans, so the row/col
+                             // is already swapped.
+    if (b_desc->_type == library_data_t::real_int8) {
+      new_b = sycl::malloc_device(sizeof(std::int8_t) * b_desc->_rows * new_ldb,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp:3051]: Can we use user-provided workspace here?

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 8]
```diff
@@ -2467,8 +2467,737 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+  dnnl::primitive_attr matmul_attr;
+  if (!vector_alpha) {
+    matmul_attr.set_scales_mask(DNNL_ARG_WEIGHTS, 0);
+    void *alpha_data = nullptr;
+    if (pointer_mode == pointer_mode_t::host) {
+      std::size_t Size =
+          dpct::detail::library_data_size[static_cast<unsigned int>(
+              scale_type)] /
+          8;
+      alpha_data = sycl::malloc_device(Size, *q_ptr);
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: alpha_data does not be free-ed.

Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: fixed

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 6 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: And use async_dpct_free

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 1 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

Comment (abhilash1910) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: If not in row/col major , should we create new instance of row/col major?
What should be the priority ? col major?

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 6 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 1 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: @abhilash1910,
For non-col-major layout input, I convert it to the col-major layout before computation.
For non-col-major layout output, I use the col-major layout to do the computation; after computation, I convert the result to the target layout.

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 9]
```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
   blas::trmm<T>(&desc, left_right, upper_lower, trans, unit_diag, m, n, alpha,
                 a, lda, b, ldb, c, ldc);
 }
+} // namespace dpct
+#ifdef __ENABLE_BLAS_EXPERIMENTAL__
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: as the imp has extra dependency on dnn lib,  and it's experimental feature,  how about put it into seperate single file?

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 10]
```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+};
+enum class pointer_mode_t {
+  host,
+  device,
+  device_vector,
+  alpha_device_vector_beta_zero,
+  alpha_device_vector_beta_host
+};
+
+class lt_handle_t;
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]:
```suggestion
class descriptor_t;   //blas_gemm_utils.hpp
```

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 11]
```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
   blas::trmm<T>(&desc, left_right, upper_lower, trans, unit_diag, m, n, alpha,
                 a, lda, b, ldb, c, ldc);
 }
+} // namespace dpct
+#ifdef __ENABLE_BLAS_EXPERIMENTAL__
+#include <oneapi/dnnl/dnnl.hpp>
+#include <oneapi/dnnl/dnnl_sycl.hpp>
+namespace dpct {
+namespace blas {
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]:
```suggestion
namespace blasgemm {
```

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 12]
```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+                               const void *alpha, const void *a,
+                               matrix_layout_ptr a_desc, const void *beta,
+                               const void *b, matrix_layout_ptr b_desc, void *c,
+                               matrix_layout_ptr c_desc, queue_ptr q_ptr);
+};
+
+class matmul_desc_t {
+public:
+  enum class attribute {
+    compute_type,
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: please clean the attr not used.

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 13]
```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+
+  friend void matmul(lt_handle_ptr handle, matmul_desc_ptr computeDesc,
+                     const void *alpha, const void *a, matrix_layout_ptr a_desc,
+                     const void *b, matrix_layout_ptr b_desc, const void *beta,
+                     const void *c, matrix_layout_ptr c_desc, void *d,
+                     matrix_layout_ptr d_desc, dpct::queue_ptr q_ptr);
+};
+
+namespace detail {
+inline dnnl::memory::data_type dpct_type_to_dnnl_type(library_data_t type) {
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: Please check if can use memory_desc_ext::to_dnnl_data_type directly.

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 14]
```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+  case library_data_t::real_uint8:
+    return dnnl::memory::data_type::u8;
+  default:
+    throw std::runtime_error(
+        "the input library_data_t type has no corresponding type in dnnl");
+  }
+}
+
+template <class T, class Talpha>
+sycl::event scale_d_with_vector_alpha_impl(queue_ptr q_ptr, int rows, int cols,
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: add comment of the algorithm.

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 15]
```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+    break;
+  }
+  default:
+    throw std::runtime_error("the combination of data type is unsupported");
+  }
+  return e;
+}
+
+inline std::tuple<size_t, size_t>
+get_from_to_linear_idx(size_t rows, size_t cols, size_t a_ld, order_t a_order,
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: // get a linear idx map for a 2D point (row_idx, col_idx) between src_order and dst_order
get_linear_idx_map(size_t rows, size_t cols, size_t src_ld, order_t src_order,
                       size_t dst_ld, order_t dst_order, size_t row_idx,
                       size_t col_idx)

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 16]
```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+        8;
+    void *beta_host = std::malloc(beta_size);
+    void *beta_zero = std::malloc(beta_size);
+    std::memset(beta_zero, 0, beta_size);
+    q_ptr->memcpy(beta_host, beta, beta_size).wait();
+    if (std::memcmp(beta_host, beta_zero, beta_size))
+      throw std::runtime_error("Non-zero beta is not supported.");
+  }
+
+  // 2. if has epilogue, need throw exception since we do not support it.
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]:
```suggestion
  // 2. if has epilogue, throw exception
```

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 17]
```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+  bool vector_alpha = false;
+  if (compute_desc->_pointer_mode == pointer_mode_t::device_vector ||
+      compute_desc->_pointer_mode ==
+          pointer_mode_t::alpha_device_vector_beta_zero ||
+      compute_desc->_pointer_mode ==
+          pointer_mode_t::alpha_device_vector_beta_host) {
+    vector_alpha = true;
+  }
+
+  // 1. if beta is not zero, need throw exception since we do not support it.
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]:
```suggestion
  // 1. if beta is not zero, need throw exception
```

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 18]
```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+  }
+  if (compute_desc->_trans_b != oneapi::mkl::transpose::trans) {
+    throw std::runtime_error("Matrix B must be transposed.");
+  }
+
+  // 3. Emit message for for non-col_major matrix.
+  if (a_desc->_order != order_t::col || b_desc->_order != order_t::col ||
+      d_desc->_order != order_t::col) {
+    // TODO: Impl row-major matmul without layout conversion
+    std::cout
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: please remove the std::out, suggest to add this limitation in function description.

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 19]
```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+  sycl::event matmul_prim_event = dnnl::sycl_interop::execute(
+      matmul_prim, handle->get_engine_stream(), matmul_args);
+  size_t dst_type_size =
+      dpct::detail::library_data_size[static_cast<unsigned int>(d_type)] / 8;
+  sycl::event e;
+  e = dpct::detail::dpct_memcpy(*q_ptr, new_d, dst_mem->get_data_handle(),
+                                dst_type_size * n * new_ldd, automatic,
+                                {matmul_prim_event});
+  if (vector_alpha)
+    e = detail::scale_d_with_vector_alpha(q_ptr, m, n, new_d, d_type, alpha,
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: 3106 has dep on e from 3102.

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 20]
```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+  else
+    dpct::async_dpct_free({alpha_data}, {e}, *q_ptr);
+  // end of calling oneDNN
+
+  if (d_desc->_order != order_t::col) {
+    if (d_desc->_type == library_data_t::real_int8) {
+      e = detail::matrix_transform<std::int8_t>(
+          q_ptr, d_desc->_rows, d_desc->_cols, new_ldd, order_t::col,
+          (const std::int8_t *)new_d, d_desc->_ld, d_desc->_order,
+          (std::int8_t *)d);
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: need explict  depend on e assigned before.

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 21]
```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+            std::tie(from_linear_idx, to_linear_idx) = get_from_to_linear_idx(
+                rows, cols, a_ld, a_order, c_ld, c_order, row_idx, col_idx);
+            c_data[to_linear_idx] = a_data[from_linear_idx];
+          }
+        });
+  });
+}
+} // namespace detail
+
+inline void matmul(lt_handle_ptr handle, matmul_desc_ptr compute_desc,
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: this API is async api, how about return the event which end user can wait on.

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp [hunk 22]
```diff
@@ -2467,8 +2467,777 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
...
+  if (transform_desc->_pointer_mode != pointer_mode_t::host) {
+    throw std::runtime_error(
+        "Only support pointer_mode_t::host as pointer_mode.");
+  }
+  if (transform_desc->_scale_type != library_data_t::real_float) {
+    throw std::runtime_error(
+        "Only support library_data_t::real_float as scale_type.");
+  }
+
+  // If alpha is not one, need throw exception since we do not support it.
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]:
```suggestion
  // If alpha is not one, need throw exception
```

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 9 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 10 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 11 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 12 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 18 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 20 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 19 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 21 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 13 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 15 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 14 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 16 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 17 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_utils.hpp (hunk 22 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_utils.hpp]: done

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 23]
```diff
@@ -0,0 +1,839 @@
+//==---- blas_gemm_utils.hpp ----------------------*- C++ -*----------------==//
+//
+// Copyright (C) Intel Corporation
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp:12]: please add comment on the purpose of this file,  diff with blas_utils.hpp file.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 24]
```diff
@@ -0,0 +1,839 @@
...
+        float_data[0] = int_data[0];
+      });
+    });
+  }
+}
+#endif
+} // namespace detail
+
+/// TODO: Impl row-major matmul without layout conversion
+/// This function does operation: D = alpha*(A*B) + beta*(C).
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp:436]: Please add more into on Type of A/B/C. 
For epilog please add todo.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 25]
```diff
@@ -0,0 +1,839 @@
...
+
+  friend sycl::event
+  matrix_transform(transform_desc_ptr transform_desc, const void *alpha,
+                   const void *a, matrix_layout_ptr a_desc, const void *beta,
+                   const void *b, matrix_layout_ptr b_desc, void *c,
+                   matrix_layout_ptr c_desc, queue_ptr q_ptr);
+};
+
+/// This function does operation:
+/// C = alpha*transformation(A) + beta*transformation(B).
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp:770]: please add more comment on transformation.

**Review by zhimingwang36 — APPROVED** [review]

**Pull request merged**